File: src/main/java/remindme/Services/TimeIntervalService.java
package remindme.Services;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.LocalTime;

import remindme.Dialogs.ManageRemind;
import remindme.Entities.TimeInterval;
import remindme.Enums.ExecutionMethod;
import remindme.Helpers.TimeRange;

public class TimeIntervalService {
    public static LocalDateTime getNextExecutionBasedOnMethod(ExecutionMethod method, TimeRange range, TimeInterval interval) {
        if (method == ExecutionMethod.CUSTOM_TIME_RANGE && ManageRemind.isTimeRangeValid(range.start(), range.end())) {
            return getNextExecutionByTimeIntervalFromSpecificTime(interval, range.start());
        }
        else if (method == ExecutionMethod.ONE_TIME_PER_DAY)  {
            return LocalDateTime.of(LocalDate.now(), range.start());
        }
        else {
            return getNextExecutionByTimeInterval(interval);
        }
    }

    public static LocalDateTime getNextExecutionByTimeInterval(TimeInterval timeInterval) {
        if (timeInterval == null) return null;

        return LocalDateTime.now().plusDays(timeInterval.days())
            .plusHours(timeInterval.hours())
            .plusMinutes(timeInterval.minutes());
    }

    public static LocalDateTime getNextExecutionByTimeIntervalFromSpecificTime(TimeInterval timeInterval, LocalTime timeFrom) {
        if (timeInterval == null || timeFrom == null) return null;

        // Base time: timeFrom
        LocalDateTime baseTime = LocalDateTime.of(LocalDate.now(), timeFrom)
            .plusDays(timeInterval.days())
            .plusHours(timeInterval.hours())
            .plusMinutes(timeInterval.minutes());

        // If the date is passed, posticipate by one day
        if (baseTime.isBefore(LocalDateTime.now())) {
            baseTime = baseTime.plusDays(1);
        }

        return baseTime;
    }
}

File: app/electron/services/timeIntervalService.ts
import type { ExecutionMethod, TimeInterval, TimeRange } from "../types";
import { addDays, combineDateAndTime, isTimeOfDayBefore } from "./dateTime";

/** Mirrors remindme.Services.TimeIntervalService. */

export function isTimeRangeValid(range: TimeRange | null): boolean {
  return range !== null && isTimeOfDayBefore(range.start, range.end);
}

export function getNextExecutionByTimeInterval(interval: TimeInterval | null, now: Date = new Date()): Date | null {
  if (!interval) return null;
  const result = new Date(now);
  result.setDate(result.getDate() + interval.days);
  result.setHours(result.getHours() + interval.hours);
  result.setMinutes(result.getMinutes() + interval.minutes);
  return result;
}

export function getNextExecutionByTimeIntervalFromSpecificTime(
  interval: TimeInterval | null,
  timeFrom: string | null,
  now: Date = new Date(),
): Date | null {
  if (!interval || !timeFrom) return null;

  let baseTime = combineDateAndTime(now, timeFrom);
  baseTime.setDate(baseTime.getDate() + interval.days);
  baseTime.setHours(baseTime.getHours() + interval.hours);
  baseTime.setMinutes(baseTime.getMinutes() + interval.minutes);

  if (baseTime < now) {
    baseTime = addDays(baseTime, 1);
  }

  return baseTime;
}

export function getNextExecutionBasedOnMethod(
  method: ExecutionMethod,
  range: TimeRange | null,
  interval: TimeInterval | null,
  now: Date = new Date(),
): Date | null {
  if (method === "CUSTOM_TIME_RANGE" && isTimeRangeValid(range)) {
    return getNextExecutionByTimeIntervalFromSpecificTime(interval, range!.start, now);
  }
  if (method === "ONE_TIME_PER_DAY" && range) {
    return combineDateAndTime(now, range.start);
  }
  return getNextExecutionByTimeInterval(interval, now);
}
